sdk/storage/azure-storage-blob/azure/storage/blob/_shared/encryption.py

```diff
@@ -115,23 +158,244 @@ def __init__(self, content_encryption_IV, encryption_agent, wrapped_content_key,
+            data_start = region_num * data_length
+            region_start = region_num * region_length
+            # Offset is based on data region
+            start_offset = start - data_start
+            # New start is the start of the encryption region
+            start = region_start
+
+        if end is not None:
+            # Find which data region the end is in
+            region_num = end // data_length
+            end_offset = start_offset + requested_length + 1
+            # New end is the end of the encryption region
+            end = (region_num * region_length) + region_length - 1
+
+    return (start, end), (start_offset, end_offset)
+
+
+def parse_encryption_data(metadata: Dict[str, Any], require_encryption: bool) -> Optional[_EncryptionData]:
+    """
+    Parses the encryption data out of the given blob metadata.
+
+    :param metadata: The blob metadata parsed from the response.
+    :param require_encryption: Whether encryption is required on the client.
+    """
+    try:
+        encryption_data_str = metadata['encryptiondata']
+        return _dict_to_encryption_data(loads(encryption_data_str))
+    except:  # pylint: disable=bare-except
+        if require_encryption:
+            raise ValueError(
+                'Encryption required, but received data does not contain appropriate metatadata.' + \
+                'Data was either not encrypted or metadata has been lost.')
+        return None
+
+
+def adjust_blob_size_for_encryption(size: int, encryption_data: Optional[_EncryptionData]) -> int:
+    """
+    Adjusts the given blob size for encryption by subtracting the size of
+    the encryption data (nonce + tag). This only has an affect for encryption V2.
+
+    :param size: The original blob size.
+    :param encryption_data: The encryption data to determine version and sizes.
+    """
+    if is_encryption_v2(encryption_data):
+        nonce_length = encryption_data.encrypted_region_info.nonce_length
+        data_length = encryption_data.encrypted_region_info.encrypted_region_data_length
+        tag_length = encryption_data.encrypted_region_info.tag_length
+        region_length = nonce_length + data_length + tag_length
+
+        num_regions = math.ceil(size / region_length)
+        metadata_size = num_regions * (nonce_length + tag_length)
+        return size - metadata_size
+
+    return size
+
+
+def _generate_encryption_data_dict(kek, cek, iv, version):
     '''
     Generates and returns the encryption metadata as a dict.
 
     :param object kek: The key encryption key. See calling functions for more information.
     :param bytes cek: The content encryption key.
-    :param bytes iv: The initialization vector.
+    :param Optional[bytes] iv: The initialization vector. Only required for AES-CBC.
+    :param str version: The client encryption version used.
     :return: A dict containing all the encryption metadata.
     :rtype: dict
```

Comment:
Nit: Just curious if API View will be happy with this `:rtype: dict` or if it would be expecting something like `:rtype: dict[Any,Any]`, but fine as-is

Reply:
This is an internal, non-exposed method. Should be fine.